### PR TITLE
Parse "Endoint=" in IoTHubEndpoint url

### PR DESCRIPTION
**Pull request by avanikp**
# Types of changes

- [X ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [ ] The code follows the code style and conventions of this project.
- [ ] The change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

# Description of the change

Env variable PCS_IOTHUBREACT_HUB_ENDPOINT does not work if "Endpoint=" is specified in the value and telemetry agent throws invalid config exception for local setup.
Fixed the issue to parse "Endpoint=" from the value if it is present.

# Motivation for the change

...

### CODE REVIEW

File: StreamingAgent/Agent.cs [hunk 1]
```diff
@@ -147,11 +148,18 @@ private void CheckConfiguration()
 
             var iotHubConnectionBuilder = IotHubConnectionStringBuilder.Create(config.IoTHubConfig.ConnectionConfig.AccessConnString);
 
-            if (!Uri.TryCreate(config.IoTHubConfig.ConnectionConfig.HubEndpoint, UriKind.Absolute, out Uri endpoint))
+            var hubEndpoint = config.IoTHubConfig.ConnectionConfig.HubEndpoint;
+            var match= Regex.Match(hubEndpoint, "^Endpoint=(?<endpoint>.*/);");
```
Comment (dluc) [inline StreamingAgent/Agent.cs]: I don't think there's a ";" at the end, is there?

Comment (dluc) [reply StreamingAgent/Agent.cs]: nit: code style, missing a space before the "=".  Please configure your IDE to automatically fix the code style "on save"

Comment (avanikp) [reply StreamingAgent/Agent.cs]: yes, there is a ';' at the end of Endpoint url (before SharedAccessKeyName). E.g "Endpoint=sb://iothub-foo-iothub-bar-12345-12345.servicebus.windows.net/;SharedAccessKeyName=iothubowner;SharedAccessKey=12345="

Comment (timlaverty) [reply StreamingAgent/Agent.cs]: I see the semicolon: AccountEndpoint=https://documentdb-host.documents.azure.com:443/;AccountKey=<my long key here>gw==;

Comment (dluc) [reply StreamingAgent/Agent.cs]: oh right, I was thinking at the full string. Shouldn't there be a `.*` after the semicolon to match, does it work anyway?

Comment (avanikp) [reply StreamingAgent/Agent.cs]: It works, we can just parse the url value which is upto ;

Comment (avanikp) [reply StreamingAgent/Agent.cs]: Added space before "="

**Comment by dluc** [comment]
Could you provide some more information explaining why we didn't encounter this issue before?

File: StreamingAgent/Agent.cs [hunk 2]
```diff
@@ -147,11 +148,18 @@ private void CheckConfiguration()
 
             var iotHubConnectionBuilder = IotHubConnectionStringBuilder.Create(config.IoTHubConfig.ConnectionConfig.AccessConnString);
 
-            if (!Uri.TryCreate(config.IoTHubConfig.ConnectionConfig.HubEndpoint, UriKind.Absolute, out Uri endpoint))
+            var hubEndpoint = config.IoTHubConfig.ConnectionConfig.HubEndpoint;
+            var match = Regex.Match(hubEndpoint, "^Endpoint=(?<endpoint>.*/);");
+            if (match.Success)
```
Comment (dluc) [inline StreamingAgent/Agent.cs]: let's move this parsing/extracting to `StreamingAgent/Runtime/Config.cs`

File: StreamingAgent/Agent.cs [hunk 3]
```diff
@@ -147,11 +148,18 @@ private void CheckConfiguration()
...
             {
-                if (!Uri.TryCreate($"sb://{config.IoTHubConfig.ConnectionConfig.HubEndpoint}/", UriKind.Absolute, out endpoint))
+                hubEndpoint = match.Groups["endpoint"].Value;
+            }
+
+            if (!Uri.TryCreate(hubEndpoint, UriKind.Absolute, out Uri endpoint))
```
Comment (dluc) [inline StreamingAgent/Agent.cs]: move validation to `StreamingAgent/Runtime/Config.cs`

File: StreamingAgent/Agent.cs (hunk 2 above)
Comment (avanikp) [reply StreamingAgent/Agent.cs]: moved

Comment (avanikp) [reply StreamingAgent/Agent.cs]: As discussed offline, I'll move CheckConfiguration() to config too.

File: StreamingAgent/Agent.cs (hunk 3 above)
Comment (avanikp) [reply StreamingAgent/Agent.cs]: done

File: StreamingAgent/Runtime/Config.cs [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
 
+using Microsoft.Azure.IoTSolutions.IoTStreamAnalytics.Services.Exceptions;
 using Microsoft.Azure.IoTSolutions.IoTStreamAnalytics.Services.Runtime;
+using System;
+using System.Text.RegularExpressions;
 
 namespace Microsoft.Azure.IoTSolutions.IoTStreamAnalytics.StreamingAgent.Runtime
 {
-    public interface IConfig
+	public interface IConfig
```
Comment (ppathan) [inline StreamingAgent/Runtime/Config.cs]: nit: spacing

File: StreamingAgent/Runtime/Config.cs [hunk 5]
```diff
@@ -134,6 +137,52 @@ public Config(IConfigData configData)
                     DocumentDbRUs = configData.GetInt(AlarmsDocDbRUsKey)
                 }
             };
-        }
+
+			CheckConfiguration(IoTHubConfig);
+		}
```
Comment (ppathan) [inline StreamingAgent/Runtime/Config.cs]: nit: spacing

File: StreamingAgent/Runtime/Config.cs [hunk 6]
```diff
@@ -134,6 +137,52 @@ public Config(IConfigData configData)
...
+				if (!Uri.TryCreate($"sb://{hubEndpoint}/", UriKind.Absolute, out endpoint))
+				{
+					{
+						throw new InvalidConfigurationException($"Invalid IoTHub endpoint {hubEndpoint}");
+					}
+				}
+			}
+
+			return hubEndpoint;
+		}
```
Comment (ppathan) [inline StreamingAgent/Runtime/Config.cs]: nit: I think extra tab is being added to code. May be want to change your IDE setting.

File: StreamingAgent.Test/ConfigTest.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+namespace StreamingAgent.Test
+{
+	public class ConfigTest
+	{
+		public ConfigTest()
+		{
+		}
+
+		[Fact]
+		public async Task HubEndpointWithEndpointPrefix()
```
Comment (ppathan) [inline StreamingAgent.Test/ConfigTest.cs]: Ideally we want to name the test that reads more like statement. Like what this test is supposed to do.

Comment (avanikp) [reply StreamingAgent.Test/ConfigTest.cs]: renamed

File: StreamingAgent/Runtime/Config.cs (hunk 4 above)
Comment (avanikp) [reply StreamingAgent/Runtime/Config.cs]: fixed

File: StreamingAgent/Runtime/Config.cs (hunk 5 above)
Comment (avanikp) [reply StreamingAgent/Runtime/Config.cs]: fixed

File: StreamingAgent.Test/ConfigTest.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		[Fact]
+		public async Task ParseHubEndpointWithEndpointPrefix()
+		{
+			// Arrange
+			Mock<IConfigData> configData = new Mock<IConfigData>();
+			configData.Setup(x => x.GetString("telemetryagent:iothub:connection:hubName")).Returns("iothub123");
+			configData.Setup(x => x.GetString("telemetryagent:iothub:checkpointing:storage:namespace")).Returns("iothub123");
+			configData.Setup(x => x.GetString("telemetryagent:iothub:checkpointing:storage:backendType")).Returns("AzureBlob");
+			configData.Setup(x => x.GetString("telemetryagent:iothub:connection:hubEndpoint"))
+				.Returns("Endpoint=sb://iothub-ns-iothub-123-1232-123.servicebus.windows.net/;SharedAccessKeyName=iothubowner;");
```
Comment (dluc) [inline StreamingAgent.Test/ConfigTest.cs]: please move this value to a constant, it's quite important for the purpose of the test

File: StreamingAgent.Test/ConfigTest.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+			// Act
+			Config config = new Config(configData.Object);
+
+			// Assert
+			Assert.Equal(config.IoTHubConfig.ConnectionConfig.HubEndpoint, "sb://iothub-ns-iothub-123-1232-123.servicebus.windows.net/");
```
Comment (dluc) [inline StreamingAgent.Test/ConfigTest.cs]: please move this value to a constant, it's quite important for the purpose of the test

File: StreamingAgent.Test/ConfigTest.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+		[Fact]
+		public async Task ParseHubEndpointWithoutEndpointPrefix()
+		{
+			// Arrange
+			var hubEndpoint = "sb://iothub-ns-iothub-123-1232-123.servicebus.windows.net/";
+			Mock<IConfigData> configData = new Mock<IConfigData>();
+			configData.Setup(x => x.GetString("telemetryagent:iothub:connection:hubName")).Returns("iothub123");
+			configData.Setup(x => x.GetString("telemetryagent:iothub:checkpointing:storage:namespace")).Returns("iothub123");
+			configData.Setup(x => x.GetString("telemetryagent:iothub:checkpointing:storage:backendType")).Returns("AzureBlob");
+			configData.Setup(x => x.GetString("telemetryagent:iothub:connection:hubEndpoint"))
```
Comment (dluc) [inline StreamingAgent.Test/ConfigTest.cs]: please move this value to a constant, it's quite important for the purpose of the test

File: StreamingAgent.Test/ConfigTest.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+				.Returns(hubEndpoint);
+
+			// Act
+			Config config = new Config(configData.Object);
+
+			// Assert
+			Assert.Equal(config.IoTHubConfig.ConnectionConfig.HubEndpoint, hubEndpoint);
```
Comment (dluc) [inline StreamingAgent.Test/ConfigTest.cs]: the order of the params is incorrect (expected/actual)

File: StreamingAgent.Test/ConfigTest.cs (hunk 9 above)
Comment (dluc) [inline StreamingAgent.Test/ConfigTest.cs]: the order of the params is incorrect (expected/actual)

File: StreamingAgent/Runtime/Config.cs [hunk 12]
```diff
@@ -134,6 +137,52 @@ public Config(IConfigData configData)
...
+            {
+                hubEndpoint = match.Groups["endpoint"].Value;
+            }
+
+            if (!Uri.TryCreate(hubEndpoint, UriKind.Absolute, out Uri endpoint))
+            {
+	    	if (!Uri.TryCreate($"sb://{hubEndpoint}/", UriKind.Absolute, out endpoint))
+	    	{
+		    {
+                        throw new InvalidConfigurationException($"Invalid IoTHub endpoint {hubEndpoint}");
```
Comment (dluc) [inline StreamingAgent/Runtime/Config.cs]: please add test to cover this expected exception

File: StreamingAgent/Runtime/Config.cs [hunk 13]
```diff
@@ -134,6 +137,52 @@ public Config(IConfigData configData)
                     DocumentDbRUs = configData.GetInt(AlarmsDocDbRUsKey)
                 }
             };
+
+	    CheckConfiguration(IoTHubConfig);
+	}
+
+        private void CheckConfiguration(IIoTHubConfig ioTHubConfig)
+        {
```
Comment (dluc) [inline StreamingAgent/Runtime/Config.cs]: please add tests to cover these expected exceptions

File: StreamingAgent.Test/ConfigTest.cs [hunk 14]
```diff
@@ -1,13 +1,14 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 
-using System.Threading.Tasks;
+using Microsoft.Azure.IoTSolutions.IoTStreamAnalytics.Services.Exceptions;
 using Microsoft.Azure.IoTSolutions.IoTStreamAnalytics.StreamingAgent.Runtime;
 using Moq;
+using System.Threading.Tasks;
```
Comment (dluc) [inline StreamingAgent.Test/ConfigTest.cs]: nit: sort/optimize imports

File: StreamingAgent.Test/ConfigTest.cs [hunk 15]
```diff
@@ -1,13 +1,14 @@
...
 using Xunit;
 
 namespace StreamingAgent.Test
 {
-	public class ConfigTest
+    public class ConfigTest
```
Comment (dluc) [inline StreamingAgent.Test/ConfigTest.cs]: code style: indentation seems off

**Pull request merged**